{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "### NOTE: could be substituted by an import!\n",
    "\n",
    "#ALPHABETS\n",
    "protein_letters = 'ACDEFGHIKLMNPQRSTVWY'\n",
    "rna_letters = 'GAUC'\n",
    "dna_letters = 'GATC'\n",
    "\n",
    "\n",
    "#WEIGHTS\n",
    "\n",
    "protein_weights = {'A': 89.09, 'C': 121.16, 'E': 147.13, 'D': 133.1, 'G': 75.07, 'F': 165.19, 'I': 131.18, 'H': 155.16, 'K': 146.19, 'M': 149.21, 'L': 131.18, 'N': 132.12, 'Q': 146.15, 'P': 115.13, 'S': 105.09, 'R': 174.2, 'T': 119.12, 'W': 204.23, 'V': 117.15, 'Y': 181.19}\n",
    "\n",
    "rna_weights = {'A': 363.0, 'C': 339.0, 'U': 340.0, 'G': 379.0}\n",
    "\n",
    "dna_weights = {'A': 347.0, 'C': 323.0, 'T': 322.0, 'G': 363.0}\n",
    "\n",
    "#COMPLEMENT\n",
    "\n",
    "dna_complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "rna_complement = {'A': 'U', 'C': 'G', 'G': 'C', 'U': 'A'}\n",
    "\n",
    "\n",
    "#CODON TABLES\n",
    "\n",
    "rna_table = {'GUC': 'V', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T', 'GUU': 'V', 'AAC': 'N', 'AGG': 'R', 'UGG': 'W', 'AGC': 'S', 'AUC': 'I', 'AGA': 'R', 'AAU': 'N', 'ACU': 'T', 'CAC': 'H', 'GUG': 'V', 'CCG': 'P', 'CCA': 'P', 'AGU': 'S', 'CCC': 'P', 'GGU': 'G', 'UCU': 'S', 'GCG': 'A', 'CGA': 'R', 'CAG': 'Q', 'CGC': 'R', 'UAU': 'Y', 'CGG': 'R', 'UCG': 'S', 'CCU': 'P', 'GGG': 'G', 'GGA': 'G', 'GGC': 'G', 'GAG': 'E', 'UCC': 'S', 'UAC': 'Y', 'CGU': 'R', 'GAA': 'E', 'AUA': 'I', 'GCA': 'A', 'CUU': 'L', 'UCA': 'S', 'AUG': 'M', 'CUG': 'L', 'AUU': 'I', 'CAU': 'H', 'CUA': 'L', 'GCC': 'A', 'AAA': 'K', 'AAG': 'K', 'CAA': 'Q', 'UUU': 'F', 'GAC': 'D', 'GUA': 'V', 'UGC': 'C', 'GCU': 'A', 'UGU': 'C', 'CUC': 'L', 'UUG': 'L', 'UUA': 'L', 'GAU': 'D', 'UUC': 'F'}\n",
    "\n",
    "rna_table_back = {'A': 'GCU', 'C': 'UGU', None: 'UAA', 'E': 'GAG', 'D': 'GAU', 'G': 'GGU', 'F': 'UUU', 'I': 'AUU', 'H': 'CAU', 'K': 'AAG', 'M': 'AUG', 'L': 'UUG', 'N': 'AAU', 'Q': 'CAG', 'P': 'CCU', 'S': 'UCU', 'R': 'CGU', 'T': 'ACU', 'W': 'UGG', 'V': 'GUU', 'Y': 'UAU'}\n",
    "\n",
    "rna_stop_codons = ['UAA', 'UAG', 'UGA']\n",
    "rna_start_codons = ['UUG', 'CUG', 'AUG']\n",
    "\n",
    "\n",
    "\n",
    "dna_table = {'CTT': 'L', 'ATG': 'M', 'ACA': 'T', 'ACG': 'T', 'ATC': 'I', 'ATA': 'I', 'AGG': 'R', 'CCT': 'P', 'AGC': 'S', 'AGA': 'R', 'ATT': 'I', 'CTG': 'L', 'CTA': 'L', 'ACT': 'T', 'CCG': 'P', 'AGT': 'S', 'CCA': 'P', 'CCC': 'P', 'TAT': 'Y', 'GGT': 'G', 'CGA': 'R', 'CGC': 'R', 'CGG': 'R', 'GGG': 'G', 'GGA': 'G', 'GGC': 'G', 'TAC': 'Y', 'CGT': 'R', 'GTA': 'V', 'GTC': 'V', 'GTG': 'V', 'GAG': 'E', 'GTT': 'V', 'GAC': 'D', 'GAA': 'E', 'AAG': 'K', 'AAA': 'K', 'AAC': 'N', 'CTC': 'L', 'CAT': 'H', 'AAT': 'N', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q', 'TGT': 'C', 'TCT': 'S', 'GAT': 'D', 'TTT': 'F', 'TGC': 'C', 'TGG': 'W', 'TTC': 'F', 'TCG': 'S', 'TTA': 'L', 'TTG': 'L', 'TCC': 'S', 'ACC': 'T', 'TCA': 'S', 'GCA': 'A', 'GCC': 'A', 'GCG': 'A', 'GCT': 'A'}\n",
    "\n",
    "dna_table_back = {'A': 'GCT', 'C': 'TGT', None: 'TAA', 'E': 'GAG', 'D': 'GAT', 'G': 'GGT', 'F': 'TTT', 'I': 'ATT', 'H': 'CAT', 'K': 'AAG', 'M': 'ATG', 'L': 'TTG', 'N': 'AAT', 'Q': 'CAG', 'P': 'CCT', 'S': 'TCT', 'R': 'CGT', 'T': 'ACT', 'W': 'TGG', 'V': 'GTT', 'Y': 'TAT'}\n",
    "\n",
    "dna_stop_codons = ['TAA', 'TAG', 'TGA']\n",
    "dna_start_codons = ['TTG', 'CTG', 'ATG']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Sequence(object):\n",
    "    alphabet = ''\n",
    "    mw = {}\n",
    "    \n",
    "    def __init__(self, identifier, sequence):\n",
    "        self.__identifier = str(identifier)\n",
    "        self.__sequence = sequence\n",
    "        \n",
    "        for monomer in self.__sequence: \n",
    "            if(monomer not in self.alphabet):\n",
    "                raise ValueError(\"Impossible to create instance: %s not possible\" %monomer)\n",
    "\n",
    "    def get_identifier(self):\n",
    "        return self.__identifier\n",
    "\n",
    "    def get_sequence(self):\n",
    "        return self.__sequence\n",
    "\n",
    "    def get_mw(self):\n",
    "        return sum(self.mw.setdefault(aa, 0) for aa in self.__sequence)\n",
    "\n",
    "    def has_subsequence(self, subsequence):\n",
    "        return subsequence.get_sequence() in self.get_sequence()\n",
    "    \n",
    "    def __len__(self): \n",
    "        return len(self.__sequence)\n",
    "    \n",
    "    def __eq__(self, other):\n",
    "        return self.get_sequence() == other.get_sequence()\n",
    "    \n",
    "    def __ne__(self, other):\n",
    "        return self.get_sequence() != other.get_sequence()\n",
    "\n",
    "    def __add__(self, other):\n",
    "        if self.__class__ != other.__class__:\n",
    "            raise ValueError(\"Operation not possible: the Sequences you are trying to add are from different classes: %s and %s\" %(str(self.__class__), str(other.__class__)))\n",
    "        sequence = str(self.get_sequence())+ str(other.get_sequence())\n",
    "        identifier = str(self.get_identifier())+\"+\"+str(other.get_identifier())\n",
    "        return self.__class__(identifier,sequence)\n",
    "    \n",
    "    def __getitem__(self, key):\n",
    "        return self.get_sequence()[key]\n",
    "    \n",
    "    def __contains__(self, item):\n",
    "        return item in self.get_sequence()\n",
    "    \n",
    "    def __lt__(self, other):\n",
    "        return self.get_mw() < other.get_mw()\n",
    "    \n",
    "    def __le__(self, other):\n",
    "        return self.get_mw() <= other.get_mw()\n",
    "    \n",
    "    def __gt__(self, other):\n",
    "        return self.get_mw() > other.get_mw()\n",
    "    \n",
    "    def __ge__(self,other):\n",
    "        return self.get_mw() >= other.get_mw()\n",
    "    \n",
    "    def __hash__(self):\n",
    "        string_to_hash = self.get_identifier()+self.get_sequence()\n",
    "        return string_to_hash.__hash__()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ProteinSequence(Sequence):\n",
    "    alphabet = protein_letters\n",
    "    mw = protein_weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "class NucleotideSequence(Sequence):\n",
    "    complement = {}\n",
    "    \n",
    "    start_codons= []\n",
    "    stop_codons = []\n",
    "    codons = {}\n",
    "    \n",
    "    def translate(self):\n",
    "        #Initialize vars\n",
    "        translation_started = False \n",
    "        translated = \"\"\n",
    "        index_beginning = 0 \n",
    "        window = 3\n",
    "        index_end = index_beginning + window\n",
    "        s = self.get_sequence()\n",
    "        \n",
    "        while index_end <= len(s):\n",
    "            index_end = index_beginning+window\n",
    "            current_codon = s[index_beginning:index_end]\n",
    "            \n",
    "            if current_codon in self.start_codons :\n",
    "                #print(\"Beginning Translation! with :\",current_codon)\n",
    "                translation_started = True\n",
    "\n",
    "            if translation_started:\n",
    "                index_beginning+= window\n",
    "                if current_codon in self.codons: \n",
    "                    translated += self.codons[current_codon]      \n",
    "                if current_codon in self.stop_codons:\n",
    "                    #print(\"Translation ended with :\",current_codon)\n",
    "                    return ProteinSequence(self.get_sequence(), translated)\n",
    "            else: \n",
    "                index_beginning+= 1\n",
    "        return ProteinSequence(self.get_sequence(), translated)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DNASequence(NucleotideSequence):\n",
    "    alphabet = dna_letters\n",
    "    mw = dna_weights\n",
    "    \n",
    "    start_codons= dna_start_codons\n",
    "    stop_codons = dna_stop_codons\n",
    "    codons = dna_table\n",
    "    \n",
    "    def transcribe(self): \n",
    "        return RNASequence(self.get_identifier(),self.get_sequence().replace(\"T\",\"U\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "class RNASequence(NucleotideSequence):\n",
    "    alphabet = rna_letters\n",
    "    mw = rna_weights\n",
    "    \n",
    "    start_codons= rna_start_codons\n",
    "    stop_codons = rna_stop_codons\n",
    "    codons = rna_table\n",
    "    \n",
    "    def reverse_transcribe(self):\n",
    "        return DNASequence(self.get_identifier(),self.get_sequence().replace(\"U\",\"T\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "# a = ProteinSequence(\"identifier1\", \"AAA\")\n",
    "# t = ProteinSequence(\"identifier1\", \"AAA\")\n",
    "# b = DNASequence(\"identifier2\", \"TAGATTGTAGTTTAGGAAAATAGAAAAAAAAAAAAAATAG\")\n",
    "# c = DNASequence(\"identifier3\", \"AAAATTTTTTTTTCC\")\n",
    "# d = RNASequence(\"identifier4\", \"UUAA\")\n",
    "# f = DNASequence(\"identifier5\", \"AAAATTCC\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
